Use additional characters as separators for stream filtering

mitanshu001 opened the pull request:
It is common to use several separators in addition to spaces between words in stream names. At present, searching in the sidebar for a word in the latter part of the stream name using such a separator will not find the stream. E.g. if a stream is called "foo-bar", searching for "bar" will not find it.

To make it easier for users to find the streams they are looking for, the left sidebar stream filter should use the following additional characters as word separators:

-, _, /

This is already being done by Zulip Terminal.

<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Fixes [#19700 ](https://github.com/zulip/zulip/issues/19700)

**Testing plan:** <!-- How have you tested? -->
Manual testing

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
<img width="131" alt="slash" src="https://user-images.githubusercontent.com/33564446/132956561-88381b02-9b3e-4d96-b59b-e9d67bc18ebe.PNG">
<img width="124" alt="space" src="https://user-images.githubusercontent.com/33564446/132956562-325007c1-218f-4025-8838-197660a91c85.PNG">
<img width="128" alt="dash" src="https://user-images.githubusercontent.com/33564446/132956565-832e8db7-37cb-4b3b-85d8-2cf9c15a44ac.PNG">
<img width="135" alt="underscore" src="https://user-images.githubusercontent.com/33564446/132956596-c8137bda-292d-449e-bb68-499c9f1fa017.PNG">

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

mitanshu001 commented:
Necessary changes are at https://github.com/zulip/zulip/pull/19726/commits/c2e1960714940f776ec2d3b1a5de4aadba579b34. Unassigning myself so I could work on another issue. 
@zulipbot abandon

timabbott commented:
@mitanshu001 thanks for working on this!  I have a few requests:
* Can you add some automated tests to the node test suite for this module?  
* What happens if `-` is present in the search query?  I'd hope that still matches properly (Would be good to have a test for this).

mitanshu001 commented:
Apologies for the late reply. Been busy for the past few days.

- I think if we are looking for a unit test then `/frontend_tests/node_tests/util.js`. `frontend_tests/node_tests/stream_sort.js` is the module that is calling the modified function.
- Do you mean let's say there is a stream `test-dash `and if the user searches for `test-` then this stream should be shown, right? or do you want this stream to be shown even if the user just searches just for `-`?

timabbott commented:
We're generally more interested in integration tests that verify correct behavior than unit tests for specific functions.  See https://zulip.readthedocs.io/en/latest/testing/philosophy.html for background.

For your second question, just that `test-` matches `test-dash`.

mitanshu001 commented:
Sorry for the late response. Was really busy for last few weeks. I have added the unit as well as the integration test. Added a test case for the `test-` as well. Please review them and let me know if anything else is required. Thanks!

alya commented:
@timabbott can this be merged?

alya commented:
Closing, as #19700 has been solved via #22253. Thanks for working on this, @mitanshu001 !
